catalyst/contrib/callbacks/knn.py

```diff
@@ -0,0 +1,218 @@
+                        postfix = self.class_names[i] \
+                            if self.class_names is not None else str(i)
+                        loader_values[f"{self.prefix}/{postfix}"] = \
+                            value
+
+        self.reset_cache()
+        self.reset_sets()
+
+    def reset_cache(self):
+        self.features = []
+        self.targets = []
+
+    def reset_sets(self):
+        self.sets = {}
+
+    def _knn(self, train_set, test_set=None):
+        """Returns accuracy calculated using kNN algorithm.
+        Args:
+            train_set: dict of feature "values" and "labels" for training set.
+            test_set: dict of feature "values" and "labels" for test set.
+        Returns:
+            cm: np.ndarray of confusion matrix
+        """
+
+        # if the test_set is None, we will test train_set on itself,
+        # in that case we need to delete the closest neighbor
+        leave_one_out = test_set is None
+
+        if leave_one_out:
+            test_set = train_set
+
+        x_train, y_train = train_set["values"], train_set["labels"]
+        x_test, y_test = test_set["values"], test_set["labels"]
+
+        size = len(y_train)
+
+        result = None
+        while result is None:
+            try:
+                y_pred = []
+
+                # fit nearest neighbors class on our train data
+                classifier = NearestNeighbors(
+                    n_neighbors=self.n_neighbors + int(leave_one_out),
+                    metric=self.knn_metric,
+                    algorithm="brute")
+                classifier.fit(x_train, y_train)
+
+                # data could be evaluated in n_folds in order to avoid OOM
+                end_idx, batch_size = 0, ceil(size / self.n_folds)
+                for s, start_idx in enumerate(range(0, size, batch_size)):
+
+                    end_idx = min(start_idx + batch_size, size)
+
+                    x = x_test[start_idx: end_idx]
+
+                    knn_ids = classifier.kneighbors(x, return_distance=False)
+
+                    # if we predict train set on itself we have to delete 0th
+                    # neighbor for all of the distances
+                    if leave_one_out:
+                        knn_ids = knn_ids[:, 1:]
+
+                    knn_classes = y_train[knn_ids]
+                    knn_classes, _ = stats.mode(knn_classes, axis=1)
+
+                    y_pred.extend(knn_classes[:, 0].tolist())
+
+                y_pred = np.asarray(y_pred)
+
+                result = (y_test, y_pred)
+
+            # this try catch block made because sometimes sets are quite big
+            # and it is not possible to put everything in memory, so we split
+            except MemoryError:
+                print(f"Memory error with {self.n_folds} fold, trying more.")
+                self.n_folds *= 2
+                result = None
+
+        return result
+
+
+__all__ = ["KNNMetricCallback"]
```

Comment:
also, please follow our Class methods organisation, like:
```
class A:
  def inner1(): ...
  def inner2(): ...
  def inner3():
    based on inner1/inner2
```

so, please place all method you use in some "main" function before this "main" function

Reply:
Done :)